src/BotPlutusInterface/Config.hs

```diff
@@ -0,0 +1,236 @@
+
+  pcChainIndexUrl <-
+    sectionWithDefault (pcChainIndexUrl def) "chainIndexUrl" ""
+
+  pcNetwork <-
+    sectionWithDefault (pcNetwork def) "networkId" ""
+
+  -- due to conflict with pcProtocolParams, should got from
+  -- pcProtocolParamsFile .json file
+  -- pcProtocolParams <-
+  --   sectionWithDefault (pcProtocolParams def) "protocolParams" ""
+  let pcProtocolParams = def
+
+  pcSlotConfig <-
+    sectionWithDefault (pcSlotConfig def) "slotConfig" ""
+
+  pcScriptFileDir <-
+    sectionWithDefault'
+      (pcScriptFileDir def)
+      "scriptFileDir"
+      pathSpec
+      "Directory name of the script and data files"
+
+  pcSigningKeyFileDir <-
+    sectionWithDefault'
+      (pcSigningKeyFileDir def)
+      "signingKeyFileDir"
+      pathSpec
+      "Directory name of the signing key files"
+
+  pcTxFileDir <-
+    sectionWithDefault'
+      (pcTxFileDir def)
+      "txFileDir"
+      pathSpec
+      "Directory name of the transaction files"
+
+  pcProtocolParamsFile <-
+    sectionWithDefault'
+      (pcProtocolParamsFile def)
+      "protocolParamsFile"
+      filepathSpec
+      "Protocol params file location relative to the cardano-cli working directory (needed for the cli) in JSON format. "
+
+  pcDryRun <-
+    sectionWithDefault'
+      (pcDryRun def)
+      "dryRun"
+      trueOrFalseSpec
+      "Dry run mode will build the tx, but skip the submit step"
+
+  pcLogLevel <-
+    sectionWithDefault' (pcLogLevel def) "logLevel" logLevelSpec ""
+
+  pcOwnPubKeyHash <-
+    sectionWithDefault (pcOwnPubKeyHash def) "ownPubKeyHash" ""
+
+  pcOwnStakePubKeyHash <-
+    sectionWithDefault' (pcOwnStakePubKeyHash def) "ownStakePubKeyHash" (maybeSpec anySpec) ""
+
+  pcTipPollingInterval <-
+    sectionWithDefault' (pcTipPollingInterval def) "tipPollingInterval" naturalSpec ""
+
+  pcForceBudget <-
+    sectionWithDefault
+      (pcForceBudget def)
+      "forceBudget"
+      "Forced budget for scripts, as optional (CPU Steps, Memory Units)"
+
+  pcPort <-
+    sectionWithDefault' (pcPort def) "port" portSpec ""
+
+  pcEnableTxEndpoint <-
+    sectionWithDefault' (pcEnableTxEndpoint def) "enableTxEndpoint" trueOrFalseSpec ""
+
+  pure PABConfig {..}
+
+docPABConfig :: String
+docPABConfig = show $ generateDocs pabConfigSpec
+
+loadPABConfig :: FilePath -> IO (Either String PABConfig)
+loadPABConfig fn = do
+  confE <- deserialize <$> readFile fn
+  case confE of
+    Left err -> return $ Left $ "PABConfig: " <> fn <> ": " <> err
+    Right conf@PABConfig {pcProtocolParamsFile} -> do
+      readProtocolParametersJSON (toString pcProtocolParamsFile)
+        <&> \case
+          Left err -> Left $ "protocolParamsFile: " <> toString pcProtocolParamsFile <> ": " <> err
+          Right pcProtocolParams -> Right conf {pcProtocolParams}
+
+savePABConfig :: FilePath -> PABConfig -> IO ()
+savePABConfig fn conf@PABConfig {pcProtocolParams, pcProtocolParamsFile} = do
+  writeProtocolParametersJSON (toString pcProtocolParamsFile) pcProtocolParams
```

Comment:
Do we even need this? Isn't `pcProtocolParams` always generated from the protocolParamsFile?

Reply:
Yes, we always use protocolParamsFile. 

I think we need this for consistency of configuration after saving. We can get it in the following ways:

1. With two functions, it can rise inconsistency problems. If the user forgot about two files, or read it from a nonstandard file name without updating protocolParamsFile and so on. 
2. By single function like now. If nobody changes `json` part, it will not be visible to a user except for access time.
3. Use Maybe for `pcProtocolParams`.

I splited this function into `savePABConfig` and `savePABConfigPartial`, mostly to highlight in API that we have two files for configuration.

Reply:
sorry @ryukzak what do you mean we need this for consistency of configuration? The protocol parameters should never be changed by the application, but even if for some reason the application does so, I don't think we should reflect those changes in the `protocol.json`.
Have a look at https://cips.cardano.org/cips/cip9/#updatableprotocolparameters changes are only made through a CIP, and will cause a hard-fork when they go through.

Reply:
Good point. I spoke about the consistency of load/save functions and between files. If we load config from two files, we need to save it in two files. 

For me, it is more about good design than practical reasons. Maybe, but I don't really expect it, we will need to make a configuration wizard.

Reply:
We need pparams for two different reasons:
- parse it, and use the data in Haskell
- store the file on disk, which is used by the cli (note, that this could be either on local or a remote machine, in the latter case we actually need two separate files)

I think it'd be nice to have some mechanism to sync the two on startup (or even better, query the cli for the current parameters if no file can be found).

In this PR however I only see this function used in tests, is there any other place I'm missing?

Reply:
And in examples.

Reply:
> I think it'd be nice to have some mechanism to sync the two on startup (or even better, query the cli for the current parameters if no file can be found).

I think this would be the best solution, the default path for that could be something like `./protocol.json` if the file is not found on startup we use the cli to fetch it and store it in whatever path there (maybe we only do this if the user has not specified a non-default path, as the chances are they might have just got the path wrong in case we don't find the file in a different path)
 
If the CLI is running on the local machine, the user must also supply the a valid `pcProtocolParamsFile` (or one will be fetched as described above) so we don't need any serialisation as we already have a file to pass to the CLI

If the CLI is running remotely I am not sure, maybe in that case we might need to serialise it to that machine before we can run the CLI from there? What assumptions do we have wrt all the files we usually have locally (signing-keys, etc) when the CLI is running remote?

Reply:
> If the CLI is running remotely I am not sure, maybe in that case we might need to serialise it to that machine before we can run the CLI from there?
yes, that would be a good approach.
> What assumptions do we have wrt all the files we usually have locally (signing-keys, etc) when the CLI is running remote?
signing-keys and protocol params must be colocated with the cli. It's an ssh connection, and the path of `pcProtocolParamsFile` and `pcSigningKeyFileDir` are relative to the home folder of the remote user

Reply:
Done, but fetching params from CLI is not checked